File: macOS/FAME Smart Flasher/Services/Flashing/FlashingService.swift
import Foundation

/// Service that orchestrates the ESP32 flashing process
actor FlashingService {
    private var connection: SerialConnection?
    private var slipDecoder = SLIPDecoder()
    private var isCancelled = false

    private let syncRetries = 20
    private let responseTimeout: TimeInterval = 5.0

    /// Flash firmware to an ESP32 device
    /// - Parameters:
    ///   - firmware: Firmware file to flash (can contain multiple images at different offsets)
    ///   - port: Serial port to use
    ///   - baudRate: Target baud rate for flashing
    ///   - onProgress: Progress callback
    func flash(
        firmware: FirmwareFile,
        port: SerialPort,
        baudRate: BaudRate,
        onProgress: @escaping (FlashingState) -> Void
    ) async throws {
        isCancelled = false
        connection = SerialConnection()

        defer {
            Task {
                await connection?.close()
            }
        }

        // 1. Connect
        await MainActor.run { onProgress(.connecting) }
        try await connection?.open(path: port.path)

        // 2. Enter bootloader mode using DTR/RTS reset sequence
        // For ESP32-C3 USB-JTAG-Serial, this triggers the built-in reset logic
        // esptool uses only one reset strategy per device type - don't mix them
        let isUSBJTAGSerial = port.isESP32C3
        #if DEBUG
        print("[FlashingService] Triggering bootloader entry via DTR/RTS (USB-JTAG-Serial: \(isUSBJTAGSerial))")
        #endif
        try await connection?.enterBootloaderMode(isUSBJTAGSerial: isUSBJTAGSerial)

        // Wait a moment for the chip to enter bootloader
        // The USB-JTAG-Serial peripheral should stay connected
        try await Task.sleep(nanoseconds: 500_000_000) // 500ms

        // Read any boot log that might have been output (like esptool does)
        // This helps diagnose whether the chip entered bootloader mode
        #if DEBUG
        if let bootData = try? await connection?.read(timeout: 0.1), !bootData.isEmpty {
            if let bootLog = String(data: bootData, encoding: .utf8) {
                print("[FlashingService] Boot output after reset: \(bootLog)")
            } else {
                print("[FlashingService] Boot output (hex): \(bootData.map { String(format: "%02X", $0) }.joined(separator: " "))")
            }
        }
        #endif

        // Flush any remaining boot messages
        await connection?.flush()

        // Try syncing without closing the port first
        // If that fails, we'll try the close/reopen approach
        var syncSucceeded = false

        #if DEBUG
        print("[FlashingService] Attempting sync without port close/reopen")
        #endif

        do {
            await MainActor.run { onProgress(.syncing) }
            try await syncWithRetry()
            syncSucceeded = true
            #if DEBUG
            print("[FlashingService] Sync succeeded without port reopen")
            #endif

            // CRITICAL: Disable watchdogs IMMEDIATELY after first sync
            // For USB-JTAG-Serial devices, the RTC watchdog can cause resets
            // that interrupt flashing. We must disable it before doing anything else.
            if isUSBJTAGSerial {
                #if DEBUG
                print("[FlashingService] Disabling watchdogs immediately after sync")
                #endif
                try await disableWatchdogs()
            }
        } catch {
            #if DEBUG
            print("[FlashingService] First sync attempt failed: \(error), trying with port reopen")
            #endif
        }

        // If sync failed, try closing and reopening the port
        // This handles cases where USB-JTAG-Serial re-enumerates
        if !syncSucceeded {
            await connection?.close()

            // Wait for USB re-enumeration
            try await Task.sleep(nanoseconds: 2_000_000_000) // 2 seconds

            // Try to reopen the port multiple times
            var opened = false
            for attempt in 1...5 {
                do {
                    try await connection?.open(path: port.path)
                    opened = true
                    #if DEBUG
                    print("[FlashingService] Port reopened on attempt \(attempt)")
                    #endif
                    break
                } catch {
                    #if DEBUG
                    print("[FlashingService] Port open attempt \(attempt) failed: \(error)")
                    #endif
                    if attempt < 5 {
                        try await Task.sleep(nanoseconds: 500_000_000) // 500ms
                    }
                }
            }

            guard opened else {
                throw FlashingError.connectionFailed("Could not reopen port after reset")
            }

            // Flush any garbage data
            await connection?.flush()

            // Try sync again
            await MainActor.run { onProgress(.syncing) }
            try await syncWithRetry()

            // CRITICAL: Disable watchdogs IMMEDIATELY after sync
            if isUSBJTAGSerial {
                #if DEBUG
                print("[FlashingService] Disabling watchdogs immediately after sync (retry path)")
                #endif
                try await disableWatchdogs()
            }
        }

        // 4. Change baud rate if needed
        if baudRate != .baud115200 {
            await MainActor.run { onProgress(.changingBaudRate) }
            try await changeBaudRate(to: baudRate)
        }

        // Note: Watchdogs already disabled immediately after sync (see above)

        // 5. Attach SPI flash (required for ROM bootloader before flash operations)
        #if DEBUG
        print("[FlashingService] Sending SPI_ATTACH command")
        #endif
        try await spiAttach()

        // 6. Flash all images in the firmware package
        let totalBytes = firmware.images.reduce(0) { $0 + $1.size }
        var bytesFlashed = 0

        #if DEBUG
        print("[FlashingService] Flashing \(firmware.images.count) image(s), total size: \(totalBytes) bytes")
        #endif

        for image in firmware.images {
            guard !isCancelled else {
                throw FlashingError.cancelled
            }

            let blockSize = ESP32Protocol.flashBlockSize
            let numBlocks = (image.size + blockSize - 1) / blockSize

            #if DEBUG
            print("[FlashingService] Flashing \(image.fileName): \(image.size) bytes at offset 0x\(String(image.offset, radix: 16))")
            let header = image.data.prefix(16)
            print("[FlashingService] Header: \(header.map { String(format: "%02X", $0) }.joined(separator: " "))")
            if image.data[0] == 0xE9 {
                print("[FlashingService] Valid ESP32 image detected (magic: 0xE9)")
            }
            #endif

            // Begin flash for this image
            await MainActor.run { onProgress(.erasing) }
            try await flashBegin(
                size: UInt32(image.size),
                numBlocks: UInt32(numBlocks),
                blockSize: UInt32(blockSize),
                offset: image.offset
            )

            // Send data blocks
            for blockNum in 0..<numBlocks {
                guard !isCancelled else {
                    throw FlashingError.cancelled
                }

                let start = blockNum * blockSize
                let end = min(start + blockSize, image.size)
                var blockData = Data(image.data[start..<end])

                // Pad last block with 0xFF if needed
                if blockData.count < blockSize {
                    blockData.append(contentsOf: [UInt8](repeating: 0xFF, count: blockSize - blockData.count))
                }

                // Calculate overall progress across all images
                let imageProgress = Double(blockNum + 1) / Double(numBlocks)
                let overallProgress = (Double(bytesFlashed) + (imageProgress * Double(image.size))) / Double(totalBytes)
                await MainActor.run { onProgress(.flashing(progress: overallProgress)) }

                try await flashData(block: blockData, sequenceNumber: blockNum)

                // Small delay after each block to prevent USB-JTAG-Serial buffer overflow
                // The ROM bootloader (without stub) can overwhelm the USB peripheral
                // This is a known issue with ESP32-C3 USB-JTAG-Serial
                try await Task.sleep(nanoseconds: 5_000_000) // 5ms between blocks
            }

            bytesFlashed += image.size
        }

        // 7. Verify (implicit - checksums validated per block)
        await MainActor.run { onProgress(.verifying) }
        try await Task.sleep(nanoseconds: 100_000_000) // Brief pause

        // 8. Complete flashing and reboot
        await MainActor.run { onProgress(.restarting) }
        try await flashEnd(reboot: true, isUSBJTAGSerial: isUSBJTAGSerial)

        try await Task.sleep(nanoseconds: 1_000_000_000) // 1 second for device to restart

        await MainActor.run { onProgress(.complete) }
    }

    /// Cancel the current flash operation
    func cancel() {
        isCancelled = true
    }

    // MARK: - Private Methods

    private func syncWithRetry() async throws {
        for attempt in 1...syncRetries {
            do {
                try await performSync()
                return // Success
            } catch {
                if attempt == syncRetries {
                    throw FlashingError.syncFailed(attempts: syncRetries)
                }
                try await Task.sleep(nanoseconds: 50_000_000) // 50ms
            }
        }
    }

    private func performSync() async throws {
        let syncCommand = ESP32Protocol.buildSyncCommand()
        let slipEncoded = SLIPCodec.encode(syncCommand)

        #if DEBUG
        print("[FlashingService] Sending SYNC command, \(slipEncoded.count) bytes")
        print("[FlashingService] SLIP encoded: \(slipEncoded.map { String(format: "%02X", $0) }.joined(separator: " "))")
        #endif

        // Send ONE sync packet (not 7 like before!)
        // esptool sends 1 sync, then reads 7 additional responses to drain
        try await connection?.write(slipEncoded)
        #if DEBUG
        print("[FlashingService] Sent sync packet")
        #endif

        // Wait for first response
        let response = try await waitForResponse(
            command: .sync,
            timeout: 1.0
        )

        #if DEBUG
        print("[FlashingService] Got response: command=\(response.command), status=\(response.status), error=\(response.error)")
        #endif

        guard response.isSuccess else {
            throw FlashingError.syncFailed(attempts: 1)
        }

        // Read 7 more responses to drain extra sync responses (like esptool does)
        // The ROM bootloader sends multiple responses to sync
        #if DEBUG
        print("[FlashingService] Draining extra sync responses...")
        #endif
        for _ in 0..<7 {
            _ = try? await waitForResponse(command: .sync, timeout: 0.1)
        }

        // Flush any remaining data
        await connection?.flush()
    }

    private func changeBaudRate(to rate: BaudRate) async throws {
        let command = ESP32Protocol.buildChangeBaudCommand(
            newBaud: UInt32(rate.rawValue),
            oldBaud: 115200
        )
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        // Brief delay then change host baud rate
        try await Task.sleep(nanoseconds: 50_000_000) // 50ms
        try await connection?.setBaudRate(rate)
        try await Task.sleep(nanoseconds: 50_000_000) // 50ms

        // Sync again at new baud rate
        try await performSync()
    }

    private func spiAttach() async throws {
        let command = ESP32Protocol.buildSpiAttachCommand()
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        let response = try await waitForResponse(command: .spiAttach, timeout: 3.0)
        #if DEBUG
        print("[FlashingService] SPI_ATTACH response: status=\(response.status), error=\(response.error)")
        #endif
        guard response.isSuccess else {
            throw FlashingError.connectionFailed("SPI attach failed: status=\(response.status), error=\(response.error)")
        }
    }

    /// Disable RTC and Super watchdogs for USB-JTAG-Serial devices
    /// This prevents the chip from resetting during long flash operations
    /// Based on esptool's watchdog handling for ESP32-C3
    private func disableWatchdogs() async throws {
        // 1. Disable RTC Watchdog
        // First unlock the write protection
        try await writeReg(
            address: ESP32C3Registers.rtcWdtWprotect,
            value: ESP32C3Registers.rtcWdtWkey
        )

        // Read current config and clear WDT_EN bit
        let wdtConfig = try await readReg(address: ESP32C3Registers.rtcWdtConfig0)
        let newWdtConfig = wdtConfig & ~ESP32C3Registers.wdtEnBit
        try await writeReg(address: ESP32C3Registers.rtcWdtConfig0, value: newWdtConfig)

        // Re-lock write protection
        try await writeReg(address: ESP32C3Registers.rtcWdtWprotect, value: 0)

        #if DEBUG
        print("[FlashingService] RTC WDT disabled (was 0x\(String(wdtConfig, radix: 16)), now 0x\(String(newWdtConfig, radix: 16)))")
        #endif

        // 2. Enable Super Watchdog auto-feed (effectively disables it)
        // First unlock the write protection
        try await writeReg(
            address: ESP32C3Registers.swdWprotect,
            value: ESP32C3Registers.swdWkey
        )

        // Read current config and set SWD_AUTO_FEED_EN bit
        let swdConfig = try await readReg(address: ESP32C3Registers.swdConf)
        let newSwdConfig = swdConfig | ESP32C3Registers.swdAutoFeedEnBit
        try await writeReg(address: ESP32C3Registers.swdConf, value: newSwdConfig)

        // Re-lock write protection
        try await writeReg(address: ESP32C3Registers.swdWprotect, value: 0)

        #if DEBUG
        print("[FlashingService] SWD auto-feed enabled (was 0x\(String(swdConfig, radix: 16)), now 0x\(String(newSwdConfig, radix: 16)))")
        #endif
    }

    private func readReg(address: UInt32) async throws -> UInt32 {
        let command = ESP32Protocol.buildReadRegCommand(address: address)
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        let response = try await waitForResponse(command: .readReg, timeout: 1.0)
        guard response.isSuccess else {
            throw FlashingError.connectionFailed("READ_REG failed at 0x\(String(address, radix: 16))")
        }
        return response.value
    }

    private func writeReg(address: UInt32, value: UInt32) async throws {
        let command = ESP32Protocol.buildWriteRegCommand(address: address, value: value)
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        let response = try await waitForResponse(command: .writeReg, timeout: 1.0)
        guard response.isSuccess else {
            throw FlashingError.connectionFailed("WRITE_REG failed at 0x\(String(address, radix: 16))")
        }
    }

    private func flashBegin(
        size: UInt32,
        numBlocks: UInt32,
        blockSize: UInt32,
        offset: UInt32
    ) async throws {
        let command = ESP32Protocol.buildFlashBeginCommand(
            size: size,
            numBlocks: numBlocks,
            blockSize: blockSize,
            offset: offset
        )
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        let response = try await waitForResponse(command: .flashBegin, timeout: 30.0) // Erase can take time
        #if DEBUG
        print("[FlashingService] FLASH_BEGIN response: status=\(response.status), error=\(response.error)")
        #endif
        guard response.isSuccess else {
            throw FlashingError.flashBeginFailed(status: response.status)
        }
    }

    private func flashData(block: Data, sequenceNumber: Int) async throws {
        let command = ESP32Protocol.buildFlashDataCommand(
            data: block,
            sequenceNumber: UInt32(sequenceNumber)
        )
        let encoded = SLIPCodec.encode(command)

        #if DEBUG
        if sequenceNumber == 0 {
            print("[FlashingService] FLASH_DATA block 0: \(encoded.count) bytes encoded, block size: \(block.count)")
            print("[FlashingService] First 60 bytes: \(encoded.prefix(60).map { String(format: "%02X", $0) }.joined(separator: " "))")
        }
        #endif

        try await connection?.write(encoded)

        let response = try await waitForResponse(command: .flashData, timeout: responseTimeout)
        #if DEBUG
        print("[FlashingService] FLASH_DATA[\(sequenceNumber)] response: status=\(response.status), error=\(response.error)")
        #endif
        guard response.isSuccess else {
            throw FlashingError.flashDataFailed(
                blockNumber: sequenceNumber,
                status: response.status
            )
        }
    }

    private func flashEnd(reboot: Bool, isUSBJTAGSerial: Bool = true) async throws {
        let command = ESP32Protocol.buildFlashEndCommand(reboot: reboot)
        let encoded = SLIPCodec.encode(command)
        try await connection?.write(encoded)

        #if DEBUG
        print("[FlashingService] Sent FLASH_END command (reboot=\(reboot))")
        #endif

        // Flash end might not get a response if rebooting
        do {
            let response = try await waitForResponse(command: .flashEnd, timeout: 2.0)
            #if DEBUG
            print("[FlashingService] FLASH_END response: status=\(response.status), error=\(response.error)")
            #endif
            if !response.isSuccess && !reboot {
                throw FlashingError.flashEndFailed
            }
        } catch {
            // Expected if rebooting
            if !reboot {
                throw error
            }
        }

        // For USB-JTAG-Serial devices, the FLASH_END reboot flag often doesn't work
        // because the ROM bootloader's soft reset doesn't reset the USB peripheral.
        // We need to do a hard reset using DTR/RTS.
        if reboot && isUSBJTAGSerial {
            #if DEBUG
            print("[FlashingService] Performing hard reset for USB-JTAG-Serial device")
            #endif
            try await connection?.hardReset()
        }
    }

    private func waitForResponse(
        command: ESP32Command,
        timeout: TimeInterval
    ) async throws -> ESP32Response {
        let deadline = Date().addingTimeInterval(timeout)
        slipDecoder.reset()

        #if DEBUG
        var totalBytesReceived = 0
        #endif

        while Date() < deadline {
            guard !isCancelled else {
                throw FlashingError.cancelled
            }

            do {
                let data = try await connection?.read(timeout: 0.1) ?? Data()

                #if DEBUG
                if !data.isEmpty {
                    totalBytesReceived += data.count
                    print("[FlashingService] Received \(data.count) bytes: \(data.prefix(50).map { String(format: "%02X", $0) }.joined(separator: " "))\(data.count > 50 ? "..." : "")")
                }
                #endif

                let packets = slipDecoder.process(data)

                for packet in packets {
                    #if DEBUG
                    print("[FlashingService] Decoded SLIP packet: \(packet.count) bytes")
                    #endif

                    if let response = ESP32Response.parse(packet),
                       response.command == command.rawValue {
                        return response
                    }
                }
            } catch SerialError.timeout {
                continue
            }
        }

        #if DEBUG
        print("[FlashingService] Timeout waiting for \(command), total bytes received: \(totalBytesReceived)")
        #endif

        throw FlashingError.timeout(operation: "waiting for \(command) response")
    }
}

File: linux/src/services/FlashingService.cpp
// FAME Smart Flasher - Linux Qt Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

#include "FlashingService.h"
#include "protocol/SLIPCodec.h"
#include "protocol/ESP32Protocol.h"

#include <QDateTime>
#include <thread>
#include <chrono>

FlashingService::FlashingService(QObject* parent)
    : QObject(parent)
{
}

FlashingService::~FlashingService()
{
    cancel();
    if (m_workerThread && m_workerThread->isRunning()) {
        m_workerThread->quit();
        m_workerThread->wait();
    }
}

void FlashingService::flash(const FirmwareFile& firmware, const SerialPort& port, BaudRate baudRate)
{
    if (m_isFlashing) {
        return;
    }

    m_isCancelled = false;
    m_isFlashing = true;

    // Run flashing in a separate thread
    m_workerThread = QThread::create([this, firmware, port, baudRate]() {
        runFlashing(firmware, port, baudRate);
    });

    connect(m_workerThread, &QThread::finished, m_workerThread, &QThread::deleteLater);
    connect(m_workerThread, &QThread::finished, this, [this]() {
        m_workerThread = nullptr;
    });

    m_workerThread->start();
}

void FlashingService::cancel()
{
    m_isCancelled = true;
}

void FlashingService::runFlashing(const FirmwareFile& firmware, const SerialPort& port, BaudRate baudRate)
{
    m_connection = std::make_unique<SerialConnection>();

    auto cleanup = [this]() {
        if (m_connection) {
            m_connection->close();
            m_connection.reset();
        }
        m_isFlashing = false;
    };

    try {
        // 1. Connect
        emit stateChanged(FlashingState::connecting());
        m_connection->open(port.path);

        // 2. Enter bootloader mode using DTR/RTS reset sequence
        // For ESP32-C3 USB-JTAG-Serial, this triggers the built-in reset logic
        // esptool uses only one reset strategy per device type - don't mix them
        bool isUSBJTAGSerial = port.isESP32C3();
        m_connection->enterBootloaderMode(isUSBJTAGSerial);

        // Wait a moment for the chip to enter bootloader
        // The USB-JTAG-Serial peripheral should stay connected
        sleepMs(500);

        // Flush any remaining boot messages
        m_connection->flush();

        // Try syncing without closing the port first
        // If that fails, we'll try the close/reopen approach
        bool syncSucceeded = false;

        try {
            emit stateChanged(FlashingState::syncing());
            syncWithRetry();
            syncSucceeded = true;

            // CRITICAL: Disable watchdogs IMMEDIATELY after first sync
            // For USB-JTAG-Serial devices, the RTC watchdog can cause resets
            // that interrupt flashing. We must disable it before doing anything else.
            if (isUSBJTAGSerial) {
                disableWatchdogs();
            }
        } catch (const std::exception&) {
            // First sync attempt failed
        }

        // If sync failed, try closing and reopening the port
        // This handles cases where USB-JTAG-Serial re-enumerates
        if (!syncSucceeded) {
            m_connection->close();

            // Wait for USB re-enumeration
            sleepMs(2000);

            // Try to reopen the port multiple times
            bool opened = false;
            for (int attempt = 1; attempt <= 5; ++attempt) {
                try {
                    m_connection->open(port.path);
                    opened = true;
                    break;
                } catch (const std::exception&) {
                    if (attempt < 5) {
                        sleepMs(500);
                    }
                }
            }

            if (!opened) {
                throw std::runtime_error("Could not reopen port after reset");
            }

            // Flush any garbage data
            m_connection->flush();

            // Try sync again
            emit stateChanged(FlashingState::syncing());
            syncWithRetry();

            // CRITICAL: Disable watchdogs IMMEDIATELY after sync
            if (isUSBJTAGSerial) {
                disableWatchdogs();
            }
        }

        // 4. Change baud rate if needed
        if (baudRate != BaudRate::Baud115200) {
            emit stateChanged(FlashingState::changingBaudRate());
            changeBaudRate(baudRate);
        }

        // 5. Attach SPI flash (required for ROM bootloader before flash operations)
        spiAttach();

        // 6. Flash all images in the firmware package
        int totalBytes = firmware.totalSize();
        int bytesFlashed = 0;

        for (const auto& image : firmware.images()) {
            if (m_isCancelled) {
                throw std::runtime_error("Cancelled");
            }

            int blockSize = ESP32Protocol::FLASH_BLOCK_SIZE;
            int numBlocks = (image.size() + blockSize - 1) / blockSize;

            // Begin flash for this image
            emit stateChanged(FlashingState::erasing());
            flashBegin(
                static_cast<uint32_t>(image.size()),
                static_cast<uint32_t>(numBlocks),
                static_cast<uint32_t>(blockSize),
                image.offset
            );

            // Send data blocks
            for (int blockNum = 0; blockNum < numBlocks; ++blockNum) {
                if (m_isCancelled) {
                    throw std::runtime_error("Cancelled");
                }

                int start = blockNum * blockSize;
                int end = qMin(start + blockSize, image.size());
                QByteArray blockData = image.data.mid(start, end - start);

                // Pad last block with 0xFF if needed
                if (blockData.size() < blockSize) {
                    blockData.append(QByteArray(blockSize - blockData.size(), static_cast<char>(0xFF)));
                }

                // Calculate overall progress across all images
                double imageProgress = static_cast<double>(blockNum + 1) / numBlocks;
                double overallProgress = (bytesFlashed + imageProgress * image.size()) / totalBytes;
                emit stateChanged(FlashingState::flashing(overallProgress));

                flashData(blockData, blockNum);

                // Small delay after each block to prevent USB-JTAG-Serial buffer overflow
                // The ROM bootloader (without stub) can overwhelm the USB peripheral
                // This is a known issue with ESP32-C3 USB-JTAG-Serial
                sleepMs(BLOCK_DELAY_MS);
            }

            bytesFlashed += image.size();
        }

        // 7. Verify (implicit - checksums validated per block)
        emit stateChanged(FlashingState::verifying());
        sleepMs(100);

        // 8. Complete flashing and reboot
        emit stateChanged(FlashingState::restarting());
        flashEnd(true, isUSBJTAGSerial);

        sleepMs(1000); // 1 second for device to restart

        emit stateChanged(FlashingState::complete());
        cleanup();
        emit finished(true);

    } catch (const std::exception& e) {
        cleanup();

        QString errorMsg = QString::fromStdString(e.what());

        if (m_isCancelled || errorMsg.contains("Cancelled")) {
            emit stateChanged(FlashingState::error(FlashingErrorType::Cancelled));
        } else if (errorMsg.contains("sync")) {
            emit stateChanged(FlashingState::error(FlashingErrorType::SyncFailed, errorMsg, SYNC_RETRIES));
        } else if (errorMsg.contains("Cannot open") || errorMsg.contains("reopen")) {
            emit stateChanged(FlashingState::error(FlashingErrorType::ConnectionFailed, errorMsg));
        } else {
            emit stateChanged(FlashingState::error(FlashingErrorType::ConnectionFailed, errorMsg));
        }

        emit finished(false);
    }
}

void FlashingService::syncWithRetry()
{
    for (int attempt = 1; attempt <= SYNC_RETRIES; ++attempt) {
        try {
            performSync();
            return; // Success
        } catch (const std::exception&) {
            if (attempt == SYNC_RETRIES) {
                throw std::runtime_error("Failed to sync with bootloader");
            }
            sleepMs(SYNC_RETRY_DELAY_MS);
        }
    }
}

void FlashingService::performSync()
{
    QByteArray syncCommand = ESP32Protocol::buildSyncCommand();
    QByteArray slipEncoded = SLIPCodec::encode(syncCommand);

    // Send ONE sync packet (not 7 like before!)
    // esptool sends 1 sync, then reads 7 additional responses to drain
    m_connection->write(slipEncoded);

    // Wait for first response
    ESP32Response response = waitForResponse(ESP32Command::Sync, 1.0);

    if (!response.isSuccess()) {
        throw std::runtime_error("Sync failed");
    }

    // Read 7 more responses to drain extra sync responses (like esptool does)
    // The ROM bootloader sends multiple responses to sync
    for (int i = 0; i < 7; ++i) {
        try {
            waitForResponse(ESP32Command::Sync, 0.1);
        } catch (...) {
            // Ignore drain timeouts
        }
    }

    // Flush any remaining data
    m_connection->flush();
}

void FlashingService::changeBaudRate(BaudRate rate)
{
    QByteArray command = ESP32Protocol::buildChangeBaudCommand(
        static_cast<uint32_t>(baudRateValue(rate)),
        115200
    );
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    // Brief delay then change host baud rate
    sleepMs(50);
    m_connection->setBaudRate(rate);
    sleepMs(50);

    // Sync again at new baud rate
    performSync();
}

void FlashingService::spiAttach()
{
    QByteArray command = ESP32Protocol::buildSpiAttachCommand();
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    ESP32Response response = waitForResponse(ESP32Command::SpiAttach, 3.0);
    if (!response.isSuccess()) {
        throw std::runtime_error(QString("SPI attach failed: status=%1, error=%2")
                                     .arg(response.status)
                                     .arg(response.error)
                                     .toStdString());
    }
}

void FlashingService::disableWatchdogs()
{
    // 1. Disable RTC Watchdog
    // First unlock the write protection
    writeReg(ESP32C3Registers::RTC_WDT_WPROTECT, ESP32C3Registers::RTC_WDT_WKEY);

    // Read current config and clear WDT_EN bit
    uint32_t wdtConfig = readReg(ESP32C3Registers::RTC_WDT_CONFIG0);
    uint32_t newWdtConfig = wdtConfig & ~ESP32C3Registers::WDT_EN_BIT;
    writeReg(ESP32C3Registers::RTC_WDT_CONFIG0, newWdtConfig);

    // Re-lock write protection
    writeReg(ESP32C3Registers::RTC_WDT_WPROTECT, 0);

    // 2. Enable Super Watchdog auto-feed (effectively disables it)
    // First unlock the write protection
    writeReg(ESP32C3Registers::SWD_WPROTECT, ESP32C3Registers::SWD_WKEY);

    // Read current config and set SWD_AUTO_FEED_EN bit
    uint32_t swdConfig = readReg(ESP32C3Registers::SWD_CONF);
    uint32_t newSwdConfig = swdConfig | ESP32C3Registers::SWD_AUTO_FEED_EN_BIT;
    writeReg(ESP32C3Registers::SWD_CONF, newSwdConfig);

    // Re-lock write protection
    writeReg(ESP32C3Registers::SWD_WPROTECT, 0);
}

uint32_t FlashingService::readReg(uint32_t address)
{
    QByteArray command = ESP32Protocol::buildReadRegCommand(address);
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    ESP32Response response = waitForResponse(ESP32Command::ReadReg, 1.0);
    if (!response.isSuccess()) {
        throw std::runtime_error(QString("READ_REG failed at 0x%1")
                                     .arg(address, 8, 16, QChar('0'))
                                     .toStdString());
    }
    return response.value;
}

void FlashingService::writeReg(uint32_t address, uint32_t value)
{
    QByteArray command = ESP32Protocol::buildWriteRegCommand(address, value);
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    ESP32Response response = waitForResponse(ESP32Command::WriteReg, 1.0);
    if (!response.isSuccess()) {
        throw std::runtime_error(QString("WRITE_REG failed at 0x%1")
                                     .arg(address, 8, 16, QChar('0'))
                                     .toStdString());
    }
}

void FlashingService::flashBegin(uint32_t size, uint32_t numBlocks, uint32_t blockSize, uint32_t offset)
{
    QByteArray command = ESP32Protocol::buildFlashBeginCommand(size, numBlocks, blockSize, offset);
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    ESP32Response response = waitForResponse(ESP32Command::FlashBegin, 30.0); // Erase can take time
    if (!response.isSuccess()) {
        throw std::runtime_error(QString("Flash begin failed: status=%1")
                                     .arg(response.status)
                                     .toStdString());
    }
}

void FlashingService::flashData(const QByteArray& block, int sequenceNumber)
{
    QByteArray command = ESP32Protocol::buildFlashDataCommand(block, static_cast<uint32_t>(sequenceNumber));
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    ESP32Response response = waitForResponse(ESP32Command::FlashData, RESPONSE_TIMEOUT);
    if (!response.isSuccess()) {
        throw std::runtime_error(QString("Flash data failed at block %1: status=%2")
                                     .arg(sequenceNumber)
                                     .arg(response.status)
                                     .toStdString());
    }
}

void FlashingService::flashEnd(bool reboot, bool isUSBJTAGSerial)
{
    QByteArray command = ESP32Protocol::buildFlashEndCommand(reboot);
    QByteArray encoded = SLIPCodec::encode(command);
    m_connection->write(encoded);

    // Flash end might not get a response if rebooting
    try {
        ESP32Response response = waitForResponse(ESP32Command::FlashEnd, 2.0);
        if (!response.isSuccess() && !reboot) {
            throw std::runtime_error("Flash end failed");
        }
    } catch (...) {
        // Expected if rebooting
        if (!reboot) {
            throw;
        }
    }

    // For USB-JTAG-Serial devices, the FLASH_END reboot flag often doesn't work
    // because the ROM bootloader's soft reset doesn't reset the USB peripheral.
    // We need to do a hard reset using DTR/RTS.
    if (reboot && isUSBJTAGSerial) {
        m_connection->hardReset();
    }
}

ESP32Response FlashingService::waitForResponse(ESP32Command command, double timeout)
{
    QDateTime deadline = QDateTime::currentDateTime().addMSecs(static_cast<qint64>(timeout * 1000));
    m_slipDecoder.reset();

    while (QDateTime::currentDateTime() < deadline) {
        if (m_isCancelled) {
            throw std::runtime_error("Cancelled");
        }

        try {
            QByteArray data = m_connection->read(0.1);

            std::vector<QByteArray> packets = m_slipDecoder.process(data);

            for (const QByteArray& packet : packets) {
                auto response = ESP32Response::parse(packet);
                if (response && response->command == static_cast<uint8_t>(command)) {
                    return *response;
                }
            }
        } catch (const SerialError& e) {
            if (e.type() == SerialError::Timeout) {
                continue;
            }
            throw;
        }
    }

    throw std::runtime_error(QString("Timeout waiting for %1 response")
                                 .arg(static_cast<int>(command))
                                 .toStdString());
}

void FlashingService::sleepMs(int ms)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(ms));
}
